Feature: openapi spec, full api browser

shamoon opened the pull request:
<!--
Note: All PRs with code changes should be targeted to the `dev` branch, pure documentation changes can target `main`
-->

## Proposed change

Going to mark as draft for now as its a big boy, but I do think its very 'safe' as this isn't affecting the app itself at all.

This:

- Implements a full openapi spec and [proper] swagger api browser
- Resolves all DRF spectacular warnings / errors and includes a new test that checks for them so if we add something that breaks it, it will tell us
- Obviously theres a ton of boilerplate stuff here, I think considering how much of an improvement this is over current we dont need to obsess that every detail is perfect but hopefully over time if there are any holes or inconsistencies etc they can be improved

<img width="1897" alt="Screenshot 2025-01-29 at 9 04 09 AM" src="https://github.com/user-attachments/assets/73ae4b49-96ba-46f4-afe0-25e6d8952c20" />
<img width="1897" alt="Screenshot 2025-01-29 at 9 04 35 AM" src="https://github.com/user-attachments/assets/bed5d58c-93a6-4bc4-bc39-3b5d55fa45b6" />


Closes #6491 (and others I think)

## Type of change

<!--
What type of change does your PR introduce to Paperless-ngx?
NOTE: Please check only one box!
-->

- [ ] Bug fix: non-breaking change which fixes an issue.
- [x] New feature / Enhancement: non-breaking change which adds functionality. _Please read the important note above._
- [ ] Breaking change: fix or feature that would cause existing functionality to not work as expected.
- [ ] Documentation only.
- [ ] Other. Please explain:

## Checklist:

<!--
NOTE: PRs that do not address the following will not be merged, please do not skip any relevant items.
-->

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [ ] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

shamoon committed:
Basic api spec changes

shamoon committed:
Fix logs error

shamoon committed:
Resolve angular dev auth warnings

shamoon committed:
Resolve some type hint warnings

shamoon committed:
Resolve obfuscated password warnings, merge use of the field

shamoon committed:
More type warnings

shamoon committed:
Api view warnings

shamoon committed:
Resolve log collision

shamoon committed:
Resolve TOTPView schema

shamoon committed:
Resolve a bunch more view warnings

shamoon committed:
Resolve the profile field

shamoon committed:
Resolve tag color warning

shamoon committed:
Resolve document filter warnings

shamoon committed:
Resolved the last logsview warning

shamoon committed:
Resolve the oauthcalllback warning

shamoon committed:
Resolve matching model enum warning

shamoon committed:
Update some settings

shamoon committed:
Better document endpoints

shamoon committed:
Better search endpoints schema

shamoon committed:
Docs start

shamoon committed:
Docs links for post, better set permissions schema

shamoon committed:
More perms stuff, some doesnt work

shamoon committed:
Bulk edit docs links

shamoon committed:
Some settings

shamoon committed:
Better test mail account

shamoon committed:
better acknowledge tasks, bulk edit ids

shamoon committed:
Make object permissions show

shamoon committed:
Fix documents permissions display

shamoon committed:
Fix operation collision

shamoon committed:
Fix some document actions but many are now wrong (not paginated)

shamoon committed:
Fix these last document actions

shamoon committed:
Manual drf-spectacular bump

shamoon committed:
Resolve nested warning on taskviewserializer

shamoon committed:
Resolve last warning, duplicate nested fields UnifiedSearchViewSet

shamoon committed:
Fix angular dev auth api schema extension

shamoon committed:
Add basic tests for schema validation

shamoon committed:
Redirect top-level API view

shamoon committed:
improve some descriptions etc

shamoon committed:
Add PaperlessBasicAuthentication schema override

shamoon committed:
Fix this

shamoon committed:
Update schema.py

shamoon committed:
typo

stumpylog reviewed (approved):
Very impressive stuff

shamoon committed:
Fix: remove exact for cf query

shamoon committed:
Fix: remove exact for cf query

Revert "Fix: remove exact for cf query"

This reverts commit 4b6b28f3004d245be2b1780d4d750e76a513dc8b.

Update custom-fields-query-dropdown.component.html

shamoon committed:
Merge branch 'dev' into feature-api-spec

shamoon committed:
Merge branch 'dev' into feature-api-spec

shamoon committed:
Merge branch 'dev' into feature-api-spec

paulgessinger commented:
This breaks (at least) my app: after login, I send a request to `/api/` with the API version that my app supports. This previously would only throw a 406 when the API version is invalid. 

The `/api` andpoint now redirects to the OpenAPI page, and that does not allow `application/json` in `Accept` any longer, and returns a 406 always.

Can we avoid having the `/api` endpoint redirect to the OpenAPI page (at least for now)? Otherwise, what's the best way to figure out now which API versions the instance supports **before logging in**?

shamoon commented:
http://localhost:8000/api/token/ seems to work fine for that purpose

```sh
curl -l -v -H "Accept: application/json;version=12" http://localhost:8000/api/token/
```

```
{"detail":"Invalid version in \"Accept\" header."}%
```

paulgessinger commented:
@shamoon I think the token endpoint only accepts POSTs (which makes sense). So in that case I'd send a POST request to `/api/token/` with the `Accept` header, and then decide if the version is fine based on the error code I get back (406 vs 405). 

Not ideal but probably workable.

shamoon commented:
It seems to respond like that even with a GET (see above), but yes either way
